lib/source/defender/aws_iot_defender_api.c

```diff
@@ -237,57 +271,58 @@ void AwsIotDefender_Stop( void )
     if( !_started )
     {
         IotLogWarn( "Defender has not started yet." );
-
-        return;
     }
-
-    /* Wait for all the metrics processing to be done, if there are outstanding requests */
-    if( IotSemaphore_GetCount( &_doneSem ) > MAX_DEFENDER_OUTSTANDING_PUBLISH_REQ )
+    else
     {
+        /* Wait for all the metrics processing to be done, if there are outstanding requests */
         IotSemaphore_Wait( &_doneSem );
-    }
 
-    IotTaskPoolJobStatus_t status;
-    IotTaskPoolError_t taskPoolError = IotTaskPool_TryCancel( IOT_SYSTEM_TASKPOOL, _metricsPublishJob, &status );
+        IotTaskPoolJobStatus_t status;
+        IotTaskPoolError_t taskPoolError = IotTaskPool_TryCancel( IOT_SYSTEM_TASKPOOL, _metricsPublishJob, &status );
 
-    /* If cancel failed, let it sleep for a while and hope everything finishes. */
-    if( taskPoolError != IOT_TASKPOOL_SUCCESS )
-    {
-        IotLogWarn( "Failed to cancel metrics publish job with return code %d and status %d.", taskPoolError, status );
-        IotClock_SleepMs( WAIT_METRICS_JOB_MAX_SECONDS * 1000 );
-    }
+        /* If cancel failed, let it sleep for a while and hope everything finishes. */
+        if( taskPoolError != IOT_TASKPOOL_SUCCESS )
+        {
+            IotLogWarn( "Failed to cancel metrics publish job with return code %d and status %d.", taskPoolError, status );
+            IotClock_SleepMs( WAIT_METRICS_JOB_MAX_SECONDS * 1000 );
+        }
+        
+        _unsubscribeMqtt();
+       
+        /* Destroy metrics' mutex. */
+        IotMutex_Destroy( &_AwsIotDefenderMetrics.mutex );
 
-    /* Destroy metrics' mutex. */
-    IotMutex_Destroy( &_AwsIotDefenderMetrics.mutex );
+        /* Post so that taken semaphore is returned */
+        IotSemaphore_Post( &_doneSem );
```

Comment:
What's the advantage of **posting** to the semaphore when it is supposed to get destroyed in the next statement?

Reply:
Destroy will fail (and trigger crash in macOS) if semaphore count is not same as initial count, which is set 1 when it this semaphore is created.